### PR TITLE
fix: can't resolve ip address with square brackets

### PR DESCRIPTION
**Pull request by digitwolf**
Problem:
Sending to destination like https://[::1]/ fails with error invalid dnsname. Error is thrown here https://github.com/rustls/hyper-rustls/blob/main/src/connector.rs#L95

hyper-rustls resolves the Hostname as hyper::Uri::host This makes hostname = "[::1]" which is HTTP-specific form of writing IPv6 address.

Solution:
Since square brakets are coming from hyper and not part standard IP notation we remove any square brakets inside of the connector before calling the rustls::ServerName::try_from(hostname)

### CODE REVIEW

File: src/connector.rs [hunk 1]
```diff
@@ -85,10 +85,16 @@ where
...
             } else if sch == &http::uri::Scheme::HTTPS {
                 let cfg = self.tls_config.clone();
-                let hostname = match self.override_server_name.as_deref() {
+                let mut hostname = match self.override_server_name.as_deref() {
                     Some(h) => h,
                     None => dst.host().unwrap_or_default(),
                 };
+
+                // Remove square brackets around IPv6 address.
+                if hostname.starts_with('[') && hostname.ends_with(']') && hostname.len() > 2 {
```
Comment (djc) [inline src/connector.rs:94]: Any reason we can't use [`trim_matches()`](https://doc.rust-lang.org/std/primitive.str.html#method.trim_matches) here?

File: Cargo.toml [hunk 2]
```diff
@@ -14,15 +14,15 @@ http = "0.2"
 hyper = { version = "0.14", default-features = false, features = ["client"] }
 log = { version = "0.4.4", optional = true }
 rustls-native-certs = { version = "0.6", optional = true }
-rustls = { version = "0.20.1", default-features = false }
+rustls = { version = "0.20.7", default-features = false }
```
Comment (djc) [inline Cargo.toml:17]: This commit doesn't help, versions in `Cargo.toml` allow the use of semantically compatible new versions by default (and in fact, Cargo will prefer newer versions), so this change is meaningless here.

File: src/connector.rs (hunk 1 above)
Comment (digitwolf) [reply src/connector.rs:94]: I wasn't aware of this Rust API. Can update this to 
```
hostname = hostname.trim_matches(|c| c == '[' || c == ']')
```

Comment (CryZe) [reply src/connector.rs:94]: That seems worse as you really only want to trim [ on the left and ] on the right.

Comment (digitwolf) [reply src/connector.rs:94]: Good point!
What about
```
hostname = hostname.trim_start_matches('[').trim_end_matches(']')
```
?

File: src/connector.rs [hunk 3]
```diff
@@ -85,10 +85,16 @@ where
...
+                    hostname = &hostname[1..hostname.len() - 1];
```
[the remark this answers is not in the document]
Comment (digitwolf) [reply src/connector.rs]: Applied your suggestion!
Thank you!

File: Cargo.toml (hunk 2 above)
Comment (djc) [reply Cargo.toml]: It's explained here: https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-cratesio.

**Review by djc — APPROVED** [review]

**Comment by digitwolf** [comment]
djc, I think I hit the wrong button, please approve the workflow again

**Comment by djc** [comment]
Okay, thanks!

**Pull request merged**